### PR TITLE
[ntuple] add support for std::optional<T>

### PR DESCRIPTION
**Pull request by jblomer**

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/RField.hxx [hunk 1]
```diff
@@ -1474,6 +1475,44 @@ public:
...
+      std::unique_ptr<RDeleter> fItemDeleter;
+
+   public:
+      explicit ROptionalDeleter(std::unique_ptr<RDeleter> itemDeleter) : fItemDeleter(std::move(itemDeleter)) {}
+      void operator()(void *objPtr, bool dtorOnly) final;
+   };
+
+   std::unique_ptr<RDeleter> fItemDeleter;
+
+   /// Given a pointer to an std::optional<T> in from, extract a pointer to the value T* and a pointer
```
Comment (enirolf) [inline tree/ntuple/v7/inc/ROOT/RField.hxx:1490]: Should this be:
```suggestion
   /// Given a pointer to an std::optional<T> in `optionalPtr`, extract a pointer to the value T* and a pointer
```
?

File: tree/ntuple/v7/test/ntuple_types.cxx [hunk 2]
```diff
@@ -1166,6 +1166,89 @@ TYPED_TEST(UniquePtr, Basics)
...
 
+TEST(RNTuple, Optional)
+{
+   FileRaii fileGuard("test_ntuple_optional.root");
+
+   {
+      auto model = RNTupleModel::Create();
+      auto pOptChar = model->MakeField<std::optional<char>>("oc");
+      auto pOptCharArr3 = model->MakeField<std::optional<std::array<char, 3>>>("oac3");
+      auto pOptCharArr4 = model->MakeField<std::optional<std::array<char, 4>>>("oac4");
```
Comment (enirolf) [inline tree/ntuple/v7/test/ntuple_types.cxx:1177]:
```suggestion
      auto pOptCharArr3 = model->MakeField<std::optional<std::array<char, 3>>>("oca3");
      auto pOptCharArr4 = model->MakeField<std::optional<std::array<char, 4>>>("oca4");
```

File: tree/ntuple/v7/test/ntuple_types.cxx [hunk 3]
```diff
@@ -1166,6 +1166,89 @@ TYPED_TEST(UniquePtr, Basics)
...
+      writer->Fill();
+   }
+
+   auto reader = RNTupleReader::Open("ntuple", fileGuard.GetPath());
+   EXPECT_EQ(3u, reader->GetNEntries());
+
+   const auto &defaultEntry = reader->GetModel().GetDefaultEntry();
+   auto pOptChar = defaultEntry.GetPtr<std::optional<char>>("oc");
+   auto pOptCharArr3 = defaultEntry.GetPtr<std::optional<std::array<char, 3>>>("oac3");
+   auto pOptCharArr4 = defaultEntry.GetPtr<std::optional<std::array<char, 4>>>("oac4");
```
Comment (enirolf) [inline tree/ntuple/v7/test/ntuple_types.cxx:1213]:
```suggestion
   auto pOptCharArr3 = defaultEntry.GetPtr<std::optional<std::array<char, 3>>>("oca3");
   auto pOptCharArr4 = defaultEntry.GetPtr<std::optional<std::array<char, 4>>>("oca4");
```

File: tree/ntuple/v7/inc/ROOT/RField.hxx [hunk 4]
```diff
@@ -1474,6 +1475,44 @@ public:
...
+   /// to the engagement boolean. Assumes that an std::optional<T> is stored as
+   /// `struct { T t; bool engagement; };`
```
Comment (enirolf) [inline tree/ntuple/v7/inc/ROOT/RField.hxx:1492]: Can there be cases where this is not true? Can/should we check for this?

**Review by enirolf — APPROVED** [review]
Looks good to me in principle! Just two nits and a question.

File: tree/ntuple/v7/inc/ROOT/RField.hxx (hunk 4 above)
Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/RField.hxx:1492]: We rely on the unit tests to catch when this is not the case. I added some more testing on the object sizes and alignments.

Comment (pcanal) [reply tree/ntuple/v7/inc/ROOT/RField.hxx:1492]: It is likely not true for the (unlikely) `std::optional<std::variant<...>>` on MacOS (i.e. it likely has the same padding issue as the converse).

File: tree/ntuple/v7/src/RField.cxx [hunk 5]
```diff
@@ -3416,6 +3422,106 @@ ROOT::Experimental::RUniquePtrField::SplitValue(const RValue &value) const
...
+   if (*engagementPtr) {
+      result.emplace_back(fSubFields[0]->BindValue(std::shared_ptr<void>(value.GetPtr<void>(), valuePtr)));
+   }
+   return result;
+}
+
+size_t ROOT::Experimental::ROptionalField::GetValueSize() const
+{
+   const auto alignment = GetAlignment();
+   const auto actualSize = fSubFields[0]->GetValueSize() + 1;
```
Comment (pcanal) [inline tree/ntuple/v7/src/RField.cxx:3508]: What's `1`? I assume it is `sizeof(char)` i.e. `sizeof(engagement)`.

Comment (pcanal) [reply tree/ntuple/v7/src/RField.cxx:3508]: or is it to make the modulo below work?

File: tree/ntuple/v7/test/ntuple_types.cxx [hunk 6]
```diff
@@ -1166,6 +1166,106 @@ TYPED_TEST(UniquePtr, Basics)
    EXPECT_EQ(nullptr, pArray->get());
 }
 
+TEST(RNTuple, Optional)
```
Comment (pcanal) [inline tree/ntuple/v7/test/ntuple_types.cxx:1169]: Are we missing test for:
```
std::optional<std::string>
std::optional<std::array<...>>
std::optional<std::map<...>>
etc.
```
and
```
std::optional<std::variant<...>>
```
?

Comment (jblomer) [reply tree/ntuple/v7/test/ntuple_types.cxx:1169]: With #15582 we have a test for `std::optional<std::variant<...>>`. I'll add test for the others.

Comment (pcanal) [reply tree/ntuple/v7/test/ntuple_types.cxx:1169]: Did we have to make adjustment to the `std::optional` support?

Comment (jblomer) [reply tree/ntuple/v7/test/ntuple_types.cxx:1169]: For `std::optional<std::variant<...>>`>? No.

Comment (jblomer) [reply tree/ntuple/v7/test/ntuple_types.cxx:1169]: I realize, `std::optional<std::array<...>>` is also already tested.

File: tree/ntuple/v7/inc/ROOT/RField.hxx (hunk 4 above)
Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/RField.hxx:1492]: It doesn't, the class inheritance structure is slightly different in this case.

File: tree/ntuple/v7/src/RField.cxx (hunk 5 above)
Comment (jblomer) [reply tree/ntuple/v7/src/RField.cxx:3508]: Clarified in https://github.com/root-project/root/pull/15720

File: tree/ntuple/v7/test/ntuple_types.cxx (hunk 6 above)
Comment (pcanal) [reply tree/ntuple/v7/test/ntuple_types.cxx:1169]: https://github.com/root-project/root/pull/15715

**Pull request merged**